slick/src/main/scala/slick/util/ManagedArrayBlockingQueue.scala

```diff
@@ -135,81 +137,64 @@ class ManagedArrayBlockingQueue[E >: Null <: PrioritizedRunnable](maximumInUse:
     }
   }
 
-  def size: Int = locked(counts)
+  // how many items in the queue
+  def size: Int = locked(itemQueue.count + highPrioItemQueue.count) // can't use `counts`
+                                                                    // here, it refers to
+                                                                    // `paused`
 
-  def remainingCapacity: Int = locked(capacity - itemQueue.count - highPrioItemQueue.count)
+  // how much normal capacity we have left before put/offer start blocking
+  def remainingCapacity: Int = math.max(locked(capacity - itemQueue.count), 0)
 
-  override def remove(o: AnyRef): Boolean = if (o eq null) false else {
-    locked {
-      if (highPrioItemQueue.remove(o)) {
-        true
-      } else {
-        itemQueue.remove(o)
-      }
+  override def remove(o: AnyRef): Boolean = if (o eq null) false else locked {
```

Comment:
`remove` is AFAICT only ever called when shutting down the AsyncExecutor, which calls `ThreadPoolExecutor.shutdownNow()` and which drains the queue.
But signalling the itemQueueNotFull condition never hurts.